src/roadmap/common.py

```diff
@@ -87,48 +88,108 @@ async def query_rbac(
     host_permissions = [p for p in permissions if p.get("permission") in inventory_access_perms]
 
     if not host_permissions:
         raise HTTPException(status_code=403, detail="Not authorized to access host inventory")
 
-    return host_permissions
+    for perm in host_permissions:
+        # Get the list of allowed Group IDs from the attribute filter.
+        for resourceDefinition in perm["resourceDefinitions"]:
+            if len(resourceDefinition) == 0:
+                # Any record with an empty resourceDefinition means
+                # unrestricted access.
+                # https://insights-rbac.readthedocs.io/en/latest/management.html#resource-definitions
+                return []
+
+            group_list = _get_group_list_from_resource_definition(resourceDefinition)
+            allowed_group_ids.update(group_list)
+
+    return allowed_group_ids
 
 
-# FIXME: This should be cached
 async def query_host_inventory(
     org_id: t.Annotated[str, Depends(decode_header)],
     session: t.Annotated[AsyncSession, Depends(get_db)],
     settings: t.Annotated[Settings, Depends(Settings.create)],
-    permissions: t.Annotated[list[dict], Depends(check_inventory_access)],
+    host_groups: t.Annotated[list[dict], Depends(get_allowed_host_groups)],
     major: MajorVersion = None,
     minor: MinorVersion = None,
 ):
     if settings.dev:
         org_id = "1234"
 
-    if any(perm.get("resourceDefinitions") for perm in permissions):
-        logger.info(f"RBAC {permissions}")
-        # TODO: Implement workspace filtering
-        raise HTTPException(501, detail="Workspace filtering is not yet implemented")
-
     query = "SELECT * FROM hbi.hosts WHERE org_id = :org_id"
     if major is not None:
         query = f"{query} AND system_profile_facts #>> '{{operating_system,major}}' = :major"
 
     if minor is not None:
         query = f"{query} AND system_profile_facts #>> '{{operating_system,minor}}' = :minor"
 
+    if host_groups:
+        # the hosts database keeps groups data in a JSONB field, with contents
+        # like this:
+        # [
+        #    {
+        #     "account": "123456",
+        #     "created": "2025-01-07T12:58:59.569065+00:00",
+        #     "host_count": 1,
+        #     "id": "f770fbf4-359d-11f0-b21b-5e43c8b8aa2f",
+        #     "name": "GroupTwo",
+        #     "org_id": "1234",
+        #     "ungrouped": false,
+        #     "updated": "2025-01-07T12:59:52.471612+00:00"
+        #    }
+        # ]
+        # This addition to the query will query into that field's list of group
+        # info. It searches for any record with an id that matches any of our
+        # eligible host group ids.
+        string_ids = [f"'{u}'" for u in host_groups]
+        id_string = ",".join(string_ids)
+        query = f"{query} AND EXISTS (SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj WHERE group_obj->>'id' IN ({id_string}))"
```

Comment:
Using f-strings to build a SQL query is asking for a SQL injection attack. The list of IDs needs to be a token and passed in as a param.

Reply:
I think SQLalchemy may do a lot of work for you.
```suggestion
        
        query = f"{query} AND EXISTS (SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj WHERE group_obj->>'id' IN :host_groups)"
    result = await session.stream(
         text(query),
         params={
            "org_id": org_id,
            "major": str(major),
            "minor": str(minor),
            "host_groups: host_groups,
        },
    )
```

Reply:
I think this query might need some work. Rather than taking the first item, we should aggregate all the `id` values from within the list of dicts in `groups`. That's a massive pain with SQL but it is possible.

I don't think grabbing one item from `groups` is correct.

Reply:
I reworked this query to use a common table expression (CTE) that creates a list of all the group IDs then only returns records where any of those values are in the list of `host_groups`.
```python
async def query_host_inventory(
    org_id: t.Annotated[str, Depends(decode_header)],
    session: t.Annotated[AsyncSession, Depends(get_db)],
    settings: t.Annotated[Settings, Depends(Settings.create)],
    host_groups: t.Annotated[list[UUID], Depends(get_allowed_host_groups)],
    major: MajorVersion = None,
    minor: MinorVersion = None,
):
    if settings.dev:
        org_id = "1234"

    query_str = "SELECT * FROM hbi.hosts WHERE org_id = :org_id"
    if host_groups:
        query_str = """
        WITH hosts_with_group_ids AS (
            SELECT *,
                (
                    SELECT ARRAY_AGG(element->>'id')
                    FROM json_array_elements(groups) AS element
                    WHERE json_array_length(groups) > 0
                ) as group_ids
            FROM hbi.hosts
            WHERE json_array_length(groups) > 0
        )
        SELECT * from hosts_with_group_ids
        WHERE group_ids && :host_groups
        """
    if major is not None:
        query_str = f"{query_str} AND system_profile_facts #>> '{{operating_system,major}}' = :major"

    if minor is not None:
        query_str = f"{query_str} AND system_profile_facts #>> '{{operating_system,minor}}' = :minor"

    query = text(textwrap.dedent(query_str))
    result = await session.stream(
        query,
        params={
            "org_id": org_id,
            "major": str(major),
            "minor": str(minor),
            "host_groups": host_groups,
        },
    )
    yield result
```

Reply:
It would probably be _much_ better to move the filters for major and minor versions inside the CTE.

Reply:
it is grabbing one item? I read it as "if the system is in any group, return 1 from the clause", which the 'exists' then checks

Reply:
I didn't see this, but I made a change and some tests. Please have a second look and tell me if we should change this. 

As to the "select 1", it's like Beav is saying, it's the same as an "any" in Python.

Reply:
I see what this is doing now. Maybe that's idiomatic SQL, but it still seems too clever and hard to follow.

IMO, it's easier to grok the final query statement if CTEs are used to gather the data we want and then query against it. That would also remove the need to inspect the query string to add the bind parameter. And if we switch to an array instead of a list of strings, we don't need an expandable bind parameter at all.

Reply:
my case for this style:

* the EXISTS (SELECT 1... style should cease searching once a match has been found. I only tested this on one client I could find, but there is a chance the speedup is greater on clients with tons of hosts and groups.  I've tried to make this efficient code understandable with indentation and comments. (the stats above show it is doing in 1.5ms what CTE style is doing in 3ms, for the one org I'm testing on.)

* the current method of chopping up/assembling the query with Python lets us have no group clauses in queries that do not use groups (probably the majority of queries, but I've not gathered any real stats). I've tried to comment this up.

* It is tricky because in certain cases we have to query for EITHER a group-id match OR a group with ungrouped = true. I think the CTE approach, if it were to be maximally readable, would require us to query for those conditions on every query not just ones where a None or group id were present, and that would be slower. (hence it's a tradeoff.)

I'm asking myself if I came back to this code with no memory and had to understand it, and I had these comments and structure, if I'd be OK with it, and I would be.

Reply:
"exists (select 1" works for me

Reply:
My overall concern with this code is legibility and maintenance, especially since it is related to access control. I don't love the solutions I have come up with. I would like to arrive at the simplest and most obvious solution, one that we will be able to modify and maintain with confidence several months from now.

Reply:
I would like to rewrite this to use the ORM because it would be easier _for me_ to read and maintain. But I'm not confident that's the best _technical_ solution.

Let's polish what we have and get comfortable with it and be confident it is doing the right thing.